packages/nimble-components/src/table/specs/data-hierarchy-hld.md

```diff
@@ -0,0 +1,168 @@
+
+### Lazy Loading
+
+In addition to just supporting data hierarchy with a fully provided set of data, we also need to support the scenario where a row is presented as a parent, but none of its children are currently represented in the data. Upon expanding the row, the client is then expected to load all of the data for that parent and provide it to the table (via the standard `setData` method).
+
+## Links To Relevant Work Items and Reference Material
+
+-   [#890: Design for hierarchical data support in tables](https://github.com/ni/nimble/issues/890)
+-   [#861: Table hierarchical data support](https://github.com/ni/nimble/issues/861)
+-   [Prototype using flat list that includes a parentId](https://github.com/ni/nimble/tree/data-hierarchy-flat-list-prototype) with [Storybook](https://60e89457a987cf003efc0a5b-qzwoshcidz.chromatic.com/?path=/story/incubating-table--table&args=data:LargeDataSet)
+-   [Prototype with user data being hierarchical](https://github.com/ni/nimble/tree/data-hierarchy-prototype) with [Storybook](https://60e89457a987cf003efc0a5b-yncupvnoes.chromatic.com/?path=/story/incubating-table--table&args=data:LargeDataSet)
+
+## Implementation / Design
+
+Users will provide hierarchical data as a flat list (just as they do for non-hierarchical data), but will have to include a field in their data that is representative of the "parent Id" for that record of data.
+
+### `Table` API
+
+```ts
+public Table() {
+    ...
+    // This attribute determines which field in the record specifies the id that is the
+    // parent row. If it is not set, the row has no parent.
+    @attr({ attribute: 'parent-id-field-name' })
+    public parentIdFieldName?: string;
+
+    // This field would be an option for how we allow a user to specify the field
+    // name where they will provide a value indicating whether or not that row
+    // is a parent row. See the below discussion on lazy-loading. If it is not set,
+    // the row is presumed child-less (and thus will show no expand/collapse button).
+    @attr({ attribute: 'has-children-field-name' })
+    public hasChildrenFieldName?: string;
+
+    // The set of rows the user would like to toggle the expand/collapse state of
+    public async expandCollapseRows(rowIds: string[]): Promise<void> {
+        ...
+    }
+}
+```
+
+### `TableRowSelectionMode` API
+
+We need the ability for a client to configure what happens when a user selects a parent row, where it could either only select the parent or would select both the parent and all of its children.
+
+```ts
+export const TableRowSelectionMode = {
+    none: undefined,
+    single: 'single',
+    multiple: 'multiple',
+    multipleWithChildren: 'multipleWithChildren' // NEW MODE!!!
```

Comment:
Some questions about this mode:
- What should happen when the user checks a row that has lazy children and then the user expands the row and the lazy children load. In the multipleWithChildren mode wouldn't the user expect those children to be checked?
- Are there examples where the operations on the parent are distinct from selection of parent + children. i.e. in a filesystem analogy selecting a parent folder and deleting it, moving it, etc, moves all the children as well without "selection" of the children.
   
   It seems like useful operations on the parent vs parent + children would be identical (the operations possible on a selection of folders + files or on states + cities). The distinguishing case would be selecting all the children, but not the parent.
   
   The question behind the question is do we actually need the multipleWithChildren mode as a distinction from the multiple mode? If so can we capture the use cases and behavior more precisely for this mode in the spec.
- Will it be common that a user just wants an easy way to select all children of an item in the hierarchy? i.e. they don't to select parent + children but just the children of a parent. If that's a common workflow, how could a user unselect the parent without unselecting all the children?